File: Baekjoon/스티커 모으기(2).cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int solution(vector<int> sticker){    
    int n = sticker.size();
    
    if(n == 1) return sticker[0];
    
    vector<vector<int>> dp(2, vector<int>(n, 0));
	dp[0][0] = 0;
	dp[1][0] = sticker[0];

	for (int i = 1; i < n; i++) {
		dp[0][i] = max(dp[0][i - 1], dp[0][i - 2] + sticker[i]);
		dp[1][i] = max(dp[1][i - 1], dp[1][i - 2] + sticker[i]);
	}
	return max(dp[0][n - 1], dp[1][n - 2]);
}

File: Programmers/스티커 모으기(2).py
def solution(sticker):
    n = len(sticker)
    if n == 1:
        return sticker[0]
    
    dp = [[0] * n for _ in range(2)]
    dp[0][0] = 0
    dp[1][0] = sticker[0]

    for i in range(2):
        for j in range(1, n):
            dp[i][j] = max(dp[i][j - 1], dp[i][j - 2] + sticker[j])
    answer = max(dp[0][-1], dp[1][-2])
    return answer